### PR TITLE
Stop wrapping action callback exceptions

### LINKED ISSUE #1640: Stop wrapping exceptions thrown by "action" callbacks in UserCallbackException
Since #1364, we wrap most exceptions thrown by user callbacks (`WhenArgumentsMatch`, `ReturnsLazily`, etc) in a `UserCallbackException`. This makes it easier to understand where the exception was thrown, especially for call specification callbacks (`A<T>.That.Matches(<callback>)`, `Where(<callback>`, `WhenArgumentsMatch(<callback>)`, etc).

We rolled back that behavior for `Invokes` to fix #1543 and now let the exception bubble up without wrapping it, because throwing from the `Invokes` callback is a legitimate scenario.

Now, #1635 makes it apparent that throwing from the `ReturnsLazily` callback is also a legitimate scenario, so we should also stop wrapping the exception in this case.

In fact, we should probably stop wrapping the exceptions thrown by all "action" callbacks:

- `Invokes` (already done)
- `ReturnsLazily`
- `AssignsOutAndRefParametersLazily`
- `Throws`

In the case of `Throws`, the callback is supposed to *return* an exception, not throw it; but it doesn't really matter if it throws, since we just throw the returned exception anyway, so the effect is the same in the end.

### PR DESCRIPTION
**Pull request by thomaslevesque**
Fixes #1640

### CODE REVIEW

File: src/FakeItEasy/Configuration/RuleBuilder.cs [hunk 1]
```diff
@@ -69,20 +69,7 @@ public IThenConfiguration<IVoidConfiguration> NumberOfTimes(int numberOfTimes)
...
-                    exceptionToThrow = exceptionFactory(call);
-                }
-                catch (Exception ex) when (!(ex is FakeConfigurationException))
-                {
-                    throw new UserCallbackException(ExceptionMessages.UserCallbackThrewAnException("Exception factory"), ex);
-                }
-
-                throw exceptionToThrow;
-            });
+            this.RuleBeingBuilt.UseApplicator(call => throw exceptionFactory(call));
```
Comment (blairconrad) [inline src/FakeItEasy/Configuration/RuleBuilder.cs]: I never noticed before, but throwing the return value does look weird. It's correct, but weird-looking!

File: src/FakeItEasy/Configuration/RuleBuilder.cs [hunk 2]
```diff
@@ -224,15 +211,7 @@ public IAfterCallConfiguredWithOutAndRefParametersConfiguration<IReturnValueConf
...
-                    try
-                    {
-                        returnValue = valueProducer(call);
-                    }
-                    catch (Exception ex) when (!(ex is FakeConfigurationException))
-                    {
-                        throw new UserCallbackException(ExceptionMessages.UserCallbackThrewAnException("Return value producer"), ex);
-                    }
-
+                    returnValue = valueProducer(call);
```
Comment (blairconrad) [inline src/FakeItEasy/Configuration/RuleBuilder.cs]: nit: snug these lines up?

**Comment by blairconrad** [comment]
Overall, very good. I have the most minor of comments. Something I expected the linter to complain about, but I see now that it didn't! I can withdraw, if you like.

File: src/FakeItEasy/Configuration/RuleBuilder.cs (hunk 1 above)
Comment (thomaslevesque) [reply src/FakeItEasy/Configuration/RuleBuilder.cs]: What's weird about it?

File: src/FakeItEasy/Configuration/RuleBuilder.cs (hunk 2 above)
Comment (thomaslevesque) [reply src/FakeItEasy/Configuration/RuleBuilder.cs]: Will do

File: src/FakeItEasy/Configuration/RuleBuilder.cs (hunk 1 above)
Comment (blairconrad) [reply src/FakeItEasy/Configuration/RuleBuilder.cs]: Just your comment earlier about how we throw the thing that's returned, but if the method throws instead, we let it go and it all works out.
Maybe it's not weird. Maybe _I'm_ weird!

File: src/FakeItEasy/Configuration/RuleBuilder.cs [hunk 3]
```diff
@@ -208,12 +208,7 @@ public IAfterCallConfiguredWithOutAndRefParametersConfiguration<IReturnValueConf
             {
                 Guard.AgainstNull(valueProducer, nameof(valueProducer));
                 this.ParentConfiguration.AddRuleIfNeeded();
-                this.ParentConfiguration.RuleBeingBuilt.UseApplicator(call =>
-                {
-                    TMember returnValue;
-                    returnValue = valueProducer(call);
-                    call.SetReturnValue(returnValue);
-                });
+                this.ParentConfiguration.RuleBeingBuilt.UseApplicator(call => call.SetReturnValue(valueProducer(call)));
```
Comment (thomaslevesque) [inline src/FakeItEasy/Configuration/RuleBuilder.cs]: Like this, or is it too much?

Comment (blairconrad) [reply src/FakeItEasy/Configuration/RuleBuilder.cs]: I'm content! The extra line with the uninitialized variable just bugged me. And probably would introduce a nullability warning! :grin:
I didn't care whether you initialized right away or elided the variable.

Comment (blairconrad) [reply src/FakeItEasy/Configuration/RuleBuilder.cs]: I was too slow to tell you I was okay with amend. Squash it and I'll merge!

**Comment by thomaslevesque** [comment]
I squashed the fixup commit.

**Comment by blairconrad** [comment]
You're fast!

**Comment by thomaslevesque** [comment]
Too fast for AppVeyor!
GitHub needs an "auto-merge" button, to merge when all checks have complete. Azure DevOps does that, it's very handy.

**Comment by blairconrad** [comment]
Exactly. I wouldn't say "no" to a "rebase with-autosquash and then merge" button ("…when all the checks are done") either!

That's a big button.

**Comment by blairconrad** [comment]
Thanks, @thomaslevesque!

**Comment by thomaslevesque** [comment]
Yeah, that'd be nice. But maybe too specific to our workflow. However, if we were able to make our own custom merge buttons... But maybe that's too much to ask for.

Thanks for the merge!

**Pull request merged**